File: app/src/main/java/org/miaowo/miaowo/custom/load_more_list/LMLPageAdapter.java
package org.miaowo.miaowo.custom.load_more_list;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * 列表
 * Created by luqin on 16-12-28.
 */

public class LMLPageAdapter<E>
        extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements LMLAdapter<E>{

    final static int NO_ID = -1;

    private List<LMLPage<E>> mPages;
    private ViewLoaderCreator<E> mLoader;

    public LMLPageAdapter(LMLPage<E> page, @NonNull ViewLoaderCreator<E> loader) {
        mPages = new ArrayList<>();
        mPages.add(page);
        mLoader = loader;
    }
    public LMLPageAdapter(List<E> items, @NonNull ViewLoaderCreator<E> loader) {
        this(new LMLPage<>(items), loader);
    }
    public LMLPageAdapter(@NonNull ViewLoaderCreator<E> loader) {
        this(new LMLPage<>(), loader);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return mLoader.createHolder(parent, viewType);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        mLoader.bindView(getItem(position), holder, getItemViewType(position));
    }

    @Override
    public int getItemCount() {
        int count = 0;
        for (LMLPage<E> item : mPages) {
            count += item.items.size();
        }
        return count;
    }
    private int getPageStart(int pageId) {
        int start = 0;
        for (LMLPage<E> page : mPages) {
            if (page.id == pageId) return start;
            else start += page.items.size();
        }
        return -1;
    }
    private int getPageUpdateCount(int pageId) {
        int after = 0;
        boolean find = false;
        for (LMLPage<E> page : mPages) {
            if (page.id == pageId) find = true;
            if (find) after += page.items.size();
        }
        return after;
    }

    @Override
    public int getItemViewType(int position) {
        return mLoader.setType(getItem(position), position);
    }

    public E getItem(int position) {
        if (mPages.size() == 0) return null;
        int pStart, pEnd = 0;
        for (LMLPage<E> page : mPages) {
            pStart = pEnd;
            pEnd += page.items.size();
            if (position < pEnd) return page.items.get(position - pStart);
        }
        return null;
    }
    public void update(List<E> newItems) {
        update(newItems, NO_ID);
    }
    public void update(List<E> newItems, int pageId) {
        if (newItems == null || newItems.isEmpty()) return;
        LMLPage<E> page = getPage(pageId);
        if (page == null) addPage(newItems, pageId);
        else {
            page.items = newItems;
            notifyItemRangeChanged(getPageStart(pageId), getPageUpdateCount(pageId));
        }
    }
    public void append(List<E> newItems, boolean toHead) {
        append(newItems, toHead, NO_ID);
    }
    public void append(List<E> newItems, boolean toHead, int pageId) {
        if (newItems == null || newItems.size() == 0) return;
        LMLPage<E> page = getPage(pageId);
        if (page == null) {
            addPage(newItems, pageId);
        } else {
            page.items.addAll(toHead ? 0 : page.items.size(), newItems);
        }
        notifyItemRangeChanged(getPageStart(pageId), getPageUpdateCount(pageId));
    }
    public void insert(E item, boolean toHead) {
        insert(item, toHead, NO_ID);
    }
    public void insert(E item, boolean toHead, int pageId) {
        LMLPage<E> page = getPage(pageId);
        if (page != null) {
            int position = toHead ? 0 : page.items.size();
            page.items.add(position, item);
            notifyItemInserted(getPageStart(pageId) + position);
        }
    }
    public void addPage(List<E> newItems, int pageId) {
        addPage(newItems, pageId, false);
    }
    public void addPage(List<E> newItems, int pageId, boolean toHead) {
        if (getPage(pageId) != null) {
            update(newItems, pageId);
            return;
        }
        mPages.add(toHead ? 0 : mPages.size(), new LMLPage<>(pageId, newItems));
        notifyItemRangeInserted(getPageStart(pageId), newItems.size());
    }
    public void removePage(int pageId) {
        LMLPage<E> page = getPage(pageId);
        if (page != null) {
            int start = getPageStart(pageId);
            mPages.remove(page);
            if (mPages.isEmpty()) mPages.add(new LMLPage<>(new ArrayList<>()));
            notifyItemRangeRemoved(start, page.items.size());
        }
    }
    public LMLPage<E> getPage(int pageId) {
        for (LMLPage<E> page : mPages) {
            if (page.id == pageId) {
                return page;
            }
        }
        return null;
    }
    public void clear() {
        if (mPages != null && !mPages.isEmpty()) {
            mPages.clear();
            notifyDataSetChanged();
        }
    }
    public boolean isIndex() {
        return mPages.get(0).index;
    }

    public interface ViewLoaderCreator<E> {
        RecyclerView.ViewHolder createHolder(ViewGroup parent, int viewType);
        void bindView(E item, RecyclerView.ViewHolder holder, int type);
        int setType(E item, int position);
    }

}

File: app/src/main/java/org/miaowo/miaowo/ui/load_more_list/LMLPageAdapter.kt
package org.miaowo.miaowo.ui.load_more_list

import org.miaowo.miaowo.other.Const
import java.util.*

/**
 * 列表
 * Created by luqin on 16-12-28.
 */

open class LMLPageAdapter<E>(page: LMLPage<E>, private val mLoader: LMLViewCreator<E>) : LMLAdapter<E>(mLoader) {
    private val mPages = ArrayList<LMLPage<E>>()
    var page = 1

    init {
        mPages.add(page)
    }

    constructor(items: List<E>, loader: LMLViewCreator<E>) : this(LMLPage(items), loader)
    constructor(loader: LMLViewCreator<E>) : this(LMLPage<E>(), loader)

    override fun getItemCount() = mPages.sumBy { it.items.size }

    private fun getPageStart(pageId: Int): Int {
        var start = 0
        for (page in mPages) {
            if (page.id == pageId)
                return start
            else
                start += page.items.size
        }
        return -1
    }

    private fun getPageUpdateCount(pageId: Int): Int {
        var after = 0
        var find = false
        for (page in mPages) {
            if (page.id == pageId) find = true
            if (find) after += page.items.size
        }
        return after
    }

    override fun getItemViewType(position: Int) = mLoader.setType(getItem(position), position)

    override fun getItem(position: Int): E? {
        if (mPages.size == 0) return null
        var pStart: Int
        var pEnd = 0
        for (page in mPages) {
            pStart = pEnd
            pEnd += page.items.size
            if (position < pEnd) return page.items[position - pStart]
        }
        return null
    }

    override fun update(newItems: List<E>) = update(newItems, Const.NO_ID)

    fun update(newItems: List<E>?, pageId: Int) {
        if (newItems == null || newItems.isEmpty()) return
        val page = getPage(pageId)
        if (page == null) addPage(newItems, pageId)
        else {
            page.items = newItems
            notifyItemRangeChanged(getPageStart(pageId), getPageUpdateCount(pageId))
        }
    }

    override fun append(newItems: List<E>, toHead: Boolean) = append(newItems, toHead, Const.NO_ID)

    fun append(newItems: List<E>?, toHead: Boolean, pageId: Int) {
        if (newItems == null || newItems.isEmpty()) return
        val list = getPage(pageId)?.items ?: mutableListOf()
        if (list is MutableList) list.addAll(if (toHead) 0 else list.size, newItems)
        notifyItemRangeChanged(getPageStart(pageId), getPageUpdateCount(pageId))
    }

    override fun insert(item: E, toHead: Boolean) = insert(item, toHead, Const.NO_ID)

    fun insert(item: E, toHead: Boolean, pageId: Int) {
        val items = getPage(pageId)?.items ?: mutableListOf()
        val position = if (toHead) 0 else items.size
        if (items is MutableList) items.add(position, item)
        notifyItemInserted(getPageStart(pageId) + position)
    }

    @JvmOverloads fun addPage(newItems: List<E>, pageId: Int, toHead: Boolean = false) {
        if (getPage(pageId) != null) {
            update(newItems, pageId)
            return
        }
        mPages.add(if (toHead) 0 else mPages.size, LMLPage(pageId, newItems))
        notifyItemRangeInserted(getPageStart(pageId), newItems.size)
    }

    fun removePage(pageId: Int) {
        val page = getPage(pageId)
        if (page != null) {
            val start = getPageStart(pageId)
            mPages.remove(page)
            if (mPages.isEmpty()) mPages.add(LMLPage(ArrayList<E>()))
            notifyItemRangeRemoved(start, page.items.size)
        }
    }

    fun getPage(pageId: Int): LMLPage<E>? = mPages.firstOrNull { it.id == pageId }

    override fun clear() {
        if (!mPages.isEmpty()) {
            mPages.clear()
            notifyDataSetChanged()
        }
    }

    override fun load(back: Boolean, list: List<E>) {
        if (back) {
            if (page > 1) page--
        }
        else page++

        if (list.isEmpty()) {
            if (!back) page--
        } else if (back) {
            if (page == 1) clear()
            addPage(list, page, true)
            removePage(page + 2)
        } else {
            addPage(list, page)
            removePage(page - 2)
        }
    }
}
